File: streams_and_processes/5_creating_a_pipe/Windows/to_upper_case.c
/*
   This program reads one character at a time from standard input,
   converts all lower case letters to upper case, and writes them
   to standard output.

   When using Windows, if standard input is the console
   keyboard, use ^z (Control-z) to denote the end of file
   (and you must use ^z at the beginning of a line!).
*/
#include <stdio.h>

int main( )
{
   int c;
   while ( (c = getchar()) != EOF )
   {
      if ( 'a' <= c && c <= 'z' )
      {
         printf("%c", c - 32);
      }
      else
      {
         printf("%c", c);
      }
   }
   return 0;
}

File: streams_and_processes/3_filter_programs/ToUpperCase.java
/*

*/

import java.util.Scanner;

/**
   This program reads lines from standard input, converts all
   the letters to upper case, and writes them to standard output.
<p>
   When using Windows, if standard input is the console
   keyboard, use ^z (Control-z) to denote the end of file
   (and you must use ^z at the beginning of a line!).

   @see ToLowerCase
*/
public class ToUpperCase
{
   public static void main(String[] args)
   {
      // Create a Scanner object to make it easier to use System.in
      final Scanner scanner = new Scanner( System.in );

      while ( scanner.hasNextLine() )
      {
         final String oneLine = scanner.nextLine();
         System.out.println( oneLine.toUpperCase() );
         if ( System.out.checkError() )
            throw new RuntimeException("System.out has encountered an IOException");
      }
   }


   // Private default constructor to enforce noninstantiable class.
   // See Item 4 in "Effective Java", 3rd Ed, Joshua Bloch.
   private ToUpperCase() {
      throw new AssertionError();
   }
}
